### ISSUE TITLE
Random disconnection of TCP connections on port 1883 (Mosquitto)

### ISSUE THREAD

**Issue by Paraphraser** [gcgarner/IOTstack#99]
### Summary

Something seems to be randomly disconnecting MQTT TCP connections on port 1883. I have no idea whether this is being caused by Mosquitto, something about the Docker container, or Raspbian. I also have no idea whether it is confined to TCP port 1883 or is more widespread.

### Discussion
 
Until I decided to adopt your absolutely brilliant IOTstack, I had been running (non-Docker) Mosquitto, Node-Red, InfluxDB and Grafana on a MacMini. All of the packages were installed and maintained via HomeBrew.

This arrangement has been reliable and stable for a good 18 months. There were several reasons why I decided to switch to IOTstack on an RPi4 but those are not important for this discussion.

The material point is:

> I have a stable reference model (the Mac-based implementation) and reasonable experience with the general behaviour of the non-Docker variants of these subsystems.

About 12 months ago I got interested in LoRa. I wound up using [WhisperNodes](https://wisen.com.au/store/products/whisper-node-lora/) as client devices and built a custom (non-TTN) LoRa Gateway based on a [Heltec ESP32 board](https://www.aliexpress.com/item/32858955542.html?spm=a2g0s.9042311.0.0.27424c4dRQZm8Z). My focus was on getting LoRa working and I have to admit that a few bugs crept into the "... and relay LoRa transmissions via MQTT" side of things.

The key bug was _not_ checking whether the [MQTTClient](https://github.com/256dpi/arduino-mqtt) instance had become disconnected. The sketch booted, brought up WiFi, established a connection with Mosquitto on the Mac and assumed that connection would stick around as long as WiFi stayed up (days..weeks).

Very slack code! But the thing is, it worked. Reliably. For over a year. Never exposing the bug. Other than a timed self-reboot every 10 days implemented in the sketch, it pretty much just ran 24x7 for a year.

Then I migrated to IOTstack running on the RPi4. Suddenly, I started to lose MQTT messages. The problem was intermittent with no obvious trigger condition but, once one MQTT packet went walkabout, nothing else would get through until I rebooted the LoRa Gateway. The LoRa Gateway has an OLED with an up-time heartbeat plus LoRa Tx/Rx packet counters. It always _looked_ like it was still working. The data just wasn't getting to Node-Red.

Took a little while (not helped by the issues raised in #91) but this was the root cause:

```
$ docker cp mosquitto:/mosquitto/log/mosquitto.log mosquitto.log

$ grep -i "Socket error on client LoRaGateway" mosquitto.log
2019-11-19T10:44:46: Socket error on client LoRaGateway, disconnecting.
2019-11-19T10:45:25: Socket error on client LoRaGateway, disconnecting.
2019-11-19T10:45:55: Socket error on client LoRaGateway, disconnecting.
2019-11-21T22:02:42: Socket error on client LoRaGateway, disconnecting.
2019-11-22T22:09:22: Socket error on client LoRaGateway, disconnecting.
2019-11-23T02:21:50: Socket error on client LoRaGateway, disconnecting.
```

The buggy code in the LoRa Gateway sketch would continue to try to send MQTT messages, totally oblivious to the fact that the TCP connection on port 1883 had disappeared.

How does a programmer react to something like this? A self-administered swift kick followed by writing the bullet-proof code that should have been written in the first place!

Since implementing the revised code on November 24, I have _not had a single example_ of a "Socket error on client LoRaGateway". I am reasonably certain that that is because of the approach I took. The LoRa Gateway only receives LoRa packets sporadically (maybe one per 15 minutes). Instead of trying to keep TCP:1883 open, the revised event loop now waits until it has something to send, opens a TCP connection on port 1883, transmits the MQTT packet, and then immediately disconnects. There is only a tiny window of opportunity for the RPi4 to interrupt this by tearing down the socket.

But that's not the end of the story because I think the foregoing has actually exposed a persistent underlying problem on the RPi4. While the words "Socket error on client" appear to lay the blame on the IoT client devices, I think that's an example of "fake news". The same client devices were _not_ having their TCP:1883 connections torn down when communicating with Mosquitto running on the Mac. The answer to "what's changed?" is "Mosquitto is now running on the RPi4". Some metrics:

```
$ docker cp mosquitto:/mosquitto/log/mosquitto.log mosquitto.log

$ head -1 mosquitto.log 
2019-11-12T12:16:26: mosquitto version 1.6.7 starting

$ tail -1 mosquitto.log 
2019-11-28T09:21:15: Client SolarRelay disconnected.

$ wc -l mosquitto.log 
11613 mosquitto.log

$ grep -i "Socket error on client" mosquitto.log|wc -l
179

$ grep -i "Socket error on client LoRaGateway" mosquitto.log|wc -l
6
```

To summarise:

* The Mosquitto log spans the better part of 16 days and some 11k entries.
* 179 of those entries are "Socket error on client". That's an arrival rate of roughly 11 per day.
* 6 of those exposed the slack code in my LoRa Gateway but, in truth, the cause of all 179 remains unexplained.

All my other IoT projects have better error checking and survive TCP:1883 going away, albeit by rebooting (which loses at least one MQTT transmission). Now that I understand the behaviour, I can improve the code in each project to reduce the probability that a random TCP:1883 teardown will derail even a single transmission (queue, reconnect, retry).

However (make that a big HOWEVER), when I compare/contrast the behaviour of non-Docker Mosquitto on the MacMini with IOTstack Mosquitto on the RPi4:

* I have looked through the last six months of logs from the Mac without finding a single instance of "Socket error on client" or anything even close to that.
* While it was communicating with Mosquitto on the Mac, the IoT device most exposed to TCP:1883 going away (the LoRa Gateway) just kept truckin'.
* Moving to the RPi4 exposed the slack code in the LoRa Gateway and, in turn, led to the discovery that the RPi4 is clobbering TCP:1883 connections at a rate around 11 per day.

Something strange is going on. Whether this is a problem in Mosquitto, or the Docker container, or the TCP stack on the RPi4 is an open question. Whatever the cause, it does not seem to be confined to ESP32/ESP8266 IoT clients. This is a breakdown of all of the "Socket error on client" messages in the Mosquitto log:

Client  | Frequency
:------------- | :-------------:
LoRaGateway  | 6
Electricity monitor  | 4
Temperature sensor	|	4
Barometric Pressure sensor	|	19
Solar monitor	|	109
\<unknown\>	|	4
mosq- prefix	|	12
mqtt_ prefix	|	21
**Total**	|	**179**

LoRaGateway is an ESP32 client. The next four are ESP8266 clients. The remainder:

* \<unknown\> appears as exactly that string. No idea.
* mosq- prefix (example "mosq-A7y9Ah7q0iuBnfFDq7"). Probably "MQTT In" nodes in the various Node-Red flows.
* mqtt_ prefix (example "mqtt_1a87050e.ce567b"). Possibly explained by my habit of monitoring MQTT traffic by running a command like this on a Mac:

```
mosquitto_sub -v -h iot-hub.local -t "/house/#" -F "%I %t %p"
``` 

I wanted to put all this on the issues table in case anyone else is chasing weird packet-loss behaviour and thinks "ahah!"

**Comment by gcgarner** [comment]
I've just pushed an update that fixed the directory permissions on the volume so now the logs should be stored correctly.

I don't know if by setting the user to 0 had an effect on mosquitto's ability control the socket. This may be a candidate for raising an issue on their project.

I've setup a test flow on nodered to send a message once every second to mosquitto. I also added a crontab to send a message every minute, hopefully the constant reconnects give it additional stress.

I will monitor the log file to see if I get any port errors.

**Comment by gcgarner** [comment]
Are you using a copper or WiFi for your network link? There was an article on hackaday where they showed that the hdmi port was causing interference on the WiFi connection and causing dropped packets on the Pi 4

**Comment by Paraphraser** [comment]
Short answer: Ethernet.

Both Ethernet & WiFi interfaces are configured and both respond to pings.

The primary name in the DNS points to the IP address that DHCP assigns to the Ethernet interface. All service names I use (eg URLs inside IoT devices for sending MQTT messages) are aliases on that primary name. To put this another way, all traffic originating on another host will be directed to the Ethernet port. That is true irrespective of whether the originating device is Ethernet or WiFi connected. My guess is that the RPi replies to traffic using the same interface the traffic arrived on so it **should** be the Ethernet port.

Netstat lists the Ethernet interface first so, unless I'm misreading that display (or Raspbian does things differently to other unices), that makes it the default interface. Yes/no?

But, having said all that, I don't actually **know** that the RPi is **not** using the WiFi interface in anger. "ifconfig" shows wlan0 packet counts are increasing. Some of those will be ordinary WiFi handshaking with the wireless hub. My Raspbian doesn't seem to have tcpdump installed so I can't have a nose around to see if it's anything more significant than that.

On the topic of the HDMI port, my RPi4 is running headless. Nothing connected to the HDMI ports. Does that make a difference?

If I were trying to answer this question on a Mac I'd just disable WiFi to see what happened. After Googling the topic for Raspbian I found plenty of "just do this then reboot" stuff, all with different flavours of "just do this". That's when I ran screaming for the hills.

If you know of a quick way to turn WiFi off and on that **doesn't** involve a reboot, I'll give it a whirl and then see if there is any reduction in disconnects in the mosquitto log.
